File: coding/src/com/algos/core/li10_tree/bt/revised/Bt_L110_BalancedBinaryTree.java
package  com.algos.core.li10_tree.bt.revised;

import com.algos.core.models.TreeNode;

public class Bt_L110_BalancedBinaryTree {

	public static void main(String[] args) {
		Bt_L110_BalancedBinaryTree obj = new Bt_L110_BalancedBinaryTree();

		TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);

		System.out.println(obj.isBalanced(root));
	}

	// Problem: https://leetcode.com/problems/balanced-binary-tree/
	// Idea: Check is height is balanced for the root node and check if left and right sub trees are balanced recursively.
	// Time: O(n)
    // space: O(1)
	public boolean isBalanced(TreeNode root) {
        return depth(root) != -1;
    }

	private int depth(TreeNode root) {
		if(root == null)
			return 0;

		int leftDepth = depth(root.left);
		int rightDepth = depth(root.right);

		if(leftDepth == -1 || rightDepth  == -1) return -1;
		if(Math.abs(leftDepth - rightDepth) > 1) return -1;

		return 1 + Math.max(leftDepth, rightDepth);
	}

}

File: dsa/concepts/10_trees/bt/python/bt_l110__balanced_binary_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/balanced-binary-tree/
# Idea: Check if the height is balanced for the root and recursively for left and right subtrees.
# Use -1 as sentinel for unbalanced subtree.
# Time: O(n)
# Space: O(1) ignoring recursion stack

def is_balanced(root):
    def depth(node):
        if not node:
            return 0

        left_depth = depth(node.left)
        if left_depth == -1:
            return -1
        right_depth = depth(node.right)
        if right_depth == -1:
            return -1

        if abs(left_depth - right_depth) > 1:
            return -1

        return 1 + max(left_depth, right_depth)

    return depth(root) != -1


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

print(is_balanced(root))  # Expected: False
